src/app/calculator/furnaces/o2-enrichment/enrichment-graph/enrichment-graph.component.ts

```diff
@@ -0,0 +1,381 @@
+    this.pointColors = graphColors;
+    this.triggerInitialResize();
+    this.initSubscriptions();
+  }
+
+  ngOnDestroy() {
+    this.enrichmentInputsSub.unsubscribe();
+  }
+
+  initSubscriptions() {
+    this.enrichmentInputsSub = this.o2EnrichmentService.enrichmentInputs.subscribe(inputs => {
+      if (inputs && this.initResizeCompleted) {
+        this.enrichmentInputs = this.o2EnrichmentService.enrichmentInputs.getValue();
+        let resetData = this.o2EnrichmentService.resetData.getValue();
+        if (this.enrichmentChart && !resetData) {
+          this.updateChart();
+        } else {
+          this.initRenderChart();
+        }
+      }
+    });
+  }
+
+  triggerInitialResize() {
+    window.dispatchEvent(new Event('resize'));
+    setTimeout(() => {
+      this.initRenderChart(true);
+    }, 25)
+  }
+
+  resizeGraph() {
+    let expandedChart = this.ngChartContainer.nativeElement;
+    if (expandedChart) {
+      if (this.expanded) {
+        this.currentChartId = this.expandedChartId;
+      }
+      else {
+        this.currentChartId = this.tabPanelChartId;
+      }
+      this.initRenderChart();
+    }
+  }
+
+  save() {
+    this.o2EnrichmentService.enrichmentChart.next(this.enrichmentChart);
+    this.o2EnrichmentService.selectedDataPoints.next(this.selectedDataPoints);
+  }
+
+  initRenderChart(isInitResize = false) {
+    Plotly.purge(this.currentChartId);
+    this.initChartSetup();
+    this.setTraces();
+
+    let chartLayout = JSON.parse(JSON.stringify(this.enrichmentChart.layout));
+    Plotly.newPlot(this.currentChartId, this.enrichmentChart.data, chartLayout, this.enrichmentChart.config)
+      .then(chart => {
+        chart.on('plotly_hover', hoverData => {
+          this.displayHoverGroupData(hoverData);
+        });
+        chart.on('plotly_unhover', unhoverData => {
+          this.removeHoverGroupData();
+        });
+      });
+    if (isInitResize == true) {
+      this.initResizeCompleted = true;
+    }
+    this.save();
+  }
+
+  initChartSetup() {
+    this.o2EnrichmentService.initChartData();
+    this.enrichmentInputs = this.o2EnrichmentService.enrichmentInputs.getValue();
+    this.enrichmentChart = this.o2EnrichmentService.enrichmentChart.getValue();
+    this.selectedDataPoints = this.o2EnrichmentService.selectedDataPoints.getValue();
+  }
+
+  updateChart() {
+    this.updateTraces();
+    let chartLayout = JSON.parse(JSON.stringify(this.enrichmentChart.layout));
+    Plotly.update(this.currentChartId, this.enrichmentChart.data, chartLayout);
+    this.save();
+  }
+
+  setTraces() {
+    this.enrichmentInputs.forEach((enrichmentInput: EnrichmentInput, index) => {
+      // Line trace
+      let graphData: TraceCoordinates = this.o2EnrichmentService.getGraphData(this.settings, enrichmentInput.inputData);
+      let lineTrace = this.o2EnrichmentService.getLineTrace();
+      lineTrace.x = graphData.x;
+      lineTrace.y = graphData.y;
+      lineTrace.line.color = this.getNextColor();
```

Comment:
Throughout this function getNextColor() is called multiple times but will retrieve the same color. Just call it once and hold it in a temp variable.